### ISSUE TITLE
"RichApi.Error Wait until the previous call completes" after executing custom function 200+ times concurrently.

### ISSUE THREAD

**Issue by jaceaser** [OfficeDev/office-js#2874]
I have a custom function which makes an asynchronous call. When I execute the function once it works fine, however if I have a worksheet with 200 of these calls and refresh the whole worksheet many of them error out with the GeneralException: Wait until the previous call completes.

## Your Environment

* Platform [PC desktop, Mac, iOS, Office on the web]: Mac
* Host [Excel, Word, PowerPoint, etc.]: Excel
* Office version number: 16.65 (22091101)
* Operating System: Monterey 12.5.1
* Browser (if using Office on the web): ______
 
## Expected behavior
I expect all the functions to execute and return the value from the webservice.

## Current behavior
Only 50 of my calls return with values the rest have exceptions.

## Steps to reproduce

1. Create a custom function with the shared runtime which makes an API call
2. Before making the API request perform an activity which requires the use of the sync with the Excel RequestContext. Like below here is my function:
/**
  * Returns the Stars
  * @customfunction 
  * @param {string} date some date.
  * @return {number} stars.
  */
async function getStarCount(date) {
    // Attempt to connect and execute queries if connection goes through
    //console.log('Executing the function')
    var context = new Excel.RequestContext();
    var formatted = context.workbook.functions.text(date, "MM/DD/YYYY");
        formatted.load("value");
        await context.sync();
        date = formatted.value
        
  const url = "https://api.github.com/repos/OfficeDev/office-js";

  let xhttp = new XMLHttpRequest();

  return new Promise(function(resolve, reject) {
    xhttp.onreadystatechange = function() {
      if (xhttp.readyState !== 4) return;

      if (xhttp.status == 200) {
        console.log(date)
        resolve(JSON.parse(xhttp.responseText).watchers_count);
      } else {
        reject({
          status: xhttp.status,

          statusText: xhttp.statusText
        });
      }
    };

    xhttp.open("GET", url, true);

    xhttp.send();
  });
}
3. Copy and paste this function in 200+ cells and hit ctrl + f9 to run all of them at once.

## Link to live example(s)

# Provide additional details
If you remove the following code:
    var context = new Excel.RequestContext();
    var formatted = context.workbook.functions.text(date, "MM/DD/YYYY");
        formatted.load("value");
        await context.sync();
        date = formatted.value

from the function it works fine.

## Context
We are unable to load complex spreadsheets which have several hundred functions. We then have to go cell by cell and reload them.

## Useful logs
[Log] Error: Wait until the previous call completes. (user-script:39, line 1)
[Log] Error Code: GeneralException (user-script:39, line 1)
[Log] Error Name: RichApi.Error (user-script:39, line 1)
GeneralException: Wait until the previous call completes.
n — excel-mac-16.00.js:25:255033
(anonymous function) — excel-mac-16.00.js:25:318873
(anonymous function) — excel-mac-16.00.js:25:317295
promiseReactionJob

Thank you for taking the time to report an issue. Our triage team will respond to you in less than 72 hours. Normally, response time is <10 hours Monday through Friday. We do not triage on weekends.

**Comment by JHJ-MS** [comment]
Hi [jaceaser](https://github.com/jaceaser), I could repro it on the Excel Desktop on Windows using the [code sample in the doc](https://learn.microsoft.com/en-us/office/dev/add-ins/excel/call-excel-apis-from-custom-function#code-sample).

I noticed [the comment on another issue](https://github.com/OfficeDev/office-js/issues/1225) by @Rick-Kirkham , 

> please take a look at this article: [Avoid using context.sync in a loop](https://docs.microsoft.com/en-us/office/dev/add-ins/concepts/correlated-objects-pattern). In particular, note this "Just as the commands in a synchronization batch job are queued, the batch jobs themselves are queued in Office, but **Office supports no more than 50 batch jobs in the queue. Any more triggers errors.**".

I wonder whether this issue is caused by the same limitation. @Rick-Kirkham, do you have any ideas about it?

**Comment by JHJ-MS** [comment]
Hi @jaceaser,  we have created an internal item 6964325 on the backlog. The root cause is still under investigation. We'll decide the priority and whether to fix it after finding the root cause. We will let you know if there are any updates. Thanks, Haijia.

**Comment by parched** [comment]
I thought I could solve this by limiting the rate in an `Excel.run` wrapper, but it still get the error approximately 25% of the time.
```
/* global Excel */
import promiseLimit from 'promise-limit';

const limit = promiseLimit<any>(45);

// Workaround for https://github.com/OfficeDev/office-js/issues/2874
// Excel only supports 50 batch jobs in the queue at a time, so we need to limit
export function runLimited<T>(batch: (context: Excel.RequestContext) => Promise<T>): Promise<T> {
  return limit(() => Excel.run(batch));
}
```
If this was the solution, I feel like this should be something `Excel.run` does by itself.

@JHJ-MS is there any way to guarantee avoiding this error other than not calling `Excel.run` from a custom function?

**Comment by parched** [comment]
Hmm, if I set the limit to 25, I can't reproduce the error. I could with 40 though.

**Comment by erbg** [comment]
Hello @parched,

can you elaborate on your solution? 

if i have a custom function like the one here: (don't look to mutch on the code it's only for demonstration)
```
export async function GetLocation(shortCode: string) {
    return await axios.get("").data;
}
```

how should i implmeent the runLimited function?

Thanks for your help

**Comment by ChemMitch** [comment]
Is there a way to tell how many items there are in the queue? 
Or a way to wait for the queue to complete the work it before adding something else to it?

**Comment by sasha-reznichenko** [comment]
years go by, the error is still there.
Is there any update on this ?

**Comment by ayush-meghwani-28** [comment]
hello @jaceaser and others who has engaged on this thread — we apologize for the lack of updates.

We will investigate based on the information provided. To help us further, a few quick follow-ups:

1. **Is this still reproducible** on your current version of Excel and Office JS? If yes, please share your platform (Windows/Mac/Web) and Office build number.
2. **Are you still actively blocked**, or have you been able to work around it? (We noted the rate-limiting approach shared by @parched — if you've found it helpful or insufficient, that context would be useful.)
3. **Has the scale changed** since you first reported — e.g. number of concurrent custom function calls, or any change in behavior with newer Office JS versions?

We will share an update on our findings as soon as we have something concrete to report.

**Comment by ayush-meghwani-28** [comment]
Hello

I reproduced this issue on my side as well (Windows Desktop Excel) using the same pattern of calling `RequestContext` + `context.sync()` from many custom-function cells at once.

## What I found

- This is a real repro.
- When many cells run in parallel, some calls fail with:
  - `RichApi.Error`
  - `GeneralException`
  - `Wait until the previous call completes`
- The Office JS batching system has a documented queue limit of about 50 batch jobs. In practice, this is why lower parallel counts (for example 40-45) often pass, and larger bursts can start failing.

## Why this happens

Each cell invocation creates its own host request and waits on `context.sync()`. During full recalculation, a large number of these can hit the host queue at the same time. Once queue pressure gets too high, Excel starts rejecting new jobs, so some cells return errors.

## Workaround to unblock now

Use a shared limiter so only a small number of Excel API calls run at once (for example 10). This prevents burst overload and avoids the queue error.

```ts
const MAX_CONCURRENT_EXCEL_REQUESTS = 10;
let activeExcelRequests = 0;
const waitingExcelRequests: Array<() => void> = [];

async function acquireExcelRequestSlot(): Promise<void> {
  if (activeExcelRequests < MAX_CONCURRENT_EXCEL_REQUESTS) {
    activeExcelRequests += 1;
    return;
  }

  await new Promise<void>((resolve) => {
    waitingExcelRequests.push(() => {
      activeExcelRequests += 1;
      resolve();
    });
  });
}

function releaseExcelRequestSlot(): void {
  activeExcelRequests -= 1;
  const next = waitingExcelRequests.shift();
  if (next) {
    next();
  }
}

async function runLimitedExcelRequest<T>(work: () => Promise<T>): Promise<T> {
  await acquireExcelRequestSlot();
  try {
    return await work();
  } finally {
    releaseExcelRequestSlot();
  }
}

// Usage inside custom function
export async function getValueThrottled(date: string): Promise<number> {
  return runLimitedExcelRequest(async () => {
    const context = new Excel.RequestContext();
    const formatted = context.workbook.functions.text(date, "MM/DD/YYYY");
    formatted.load("value");
    await context.sync();
    return String(formatted.value).length;
  });
}
```

This workaround usually trades some speed for stability, which is better than intermittent `#VALUE!` failures in large sheets.
